File: src/sugtao4423/yandereviewer/PostDetail.java
package sugtao4423.yandereviewer;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.DecimalFormat;
import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import android.app.ActionBar;
import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.Html;
import android.text.format.DateFormat;
import android.text.method.LinkMovementMethod;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;
import yandere4j.Yandere4j;
import yandere4j.data.Post;

public class PostDetail extends Activity{

	private Post post;
	private boolean onIntent;

	private TextView text;

	@Override
	protected void onCreate(Bundle savedInstanceState){
		super.onCreate(savedInstanceState);
		setContentView(R.layout.post_detail);
		text = (TextView)findViewById(R.id.postDetail_text);
		post = (Post)getIntent().getSerializableExtra("postdata");
		onIntent = getIntent().getBooleanExtra("onIntent", false);

		String tags = "";
		for(String s : post.getTags())
			tags += s + " ";
		tags = tags.substring(0, tags.length() - 1);

		getActionBar().setTitle(tags);
		setActionbarIcon();

		String source;
		Matcher urlMatcher = Pattern.compile("http(s)?://([\\w-]+\\.)+[\\w-]+(/[\\w- ./?%&=]*)?").matcher(post.getSource());
		if(urlMatcher.find()){
			Matcher pixivMatcher = Pattern.compile(
					"http(s)?://(i[0-9].pixiv|i.pximg).net/img-original/img/[0-9]{4}/([0-9]{2}/){5}([0-9]+)_p[0-9]+\\..+")
					.matcher(post.getSource());
			source = pixivMatcher.find() ?
					"http://www.pixiv.net/member_illust.php?mode=medium&illust_id=" + pixivMatcher.group(4) : post.getSource();
		}else{
			source = "http://google.com/search?q=" + post.getSource();
		}
		String date = DateFormat.getDateFormat(getApplicationContext()).format(post.getCreatedAt());

		String str = "<p><strong>Statistics</strong><br />" +
				"Id: " + post.getId() + "<br />" +
				"Posted: " + date + " " + new SimpleDateFormat("HH:mm:ss", Locale.getDefault()).format(post.getCreatedAt()) +
							" by " + post.getAuthor() + "<br />" +
				"Tags: " + tags + "<br />" +
				"Size: " + post.getFile().getWidth() + "x" + post.getFile().getHeight() + "<br />" +
				"Source: <a href=\"" + source + "\">" + post.getSource() + "</a><br />" +
				"Rating: " + post.getRating() + "<br />" +
				"Score: " + post.getScore() + "<br /><br />" +
				"<strong>Preview</strong><br />" +
				"URL: <a href=\"" + post.getPreview().getUrl() + "\">" + post.getPreview().getUrl() + "</a><br />" +
				"Size: " + post.getPreview().getWidth() + "x" + post.getPreview().getHeight() + "<br /><br />" +
				"<strong>Sample</strong><br />" +
				"URL: <a href=\"" + post.getSample().getUrl() + "\">" + post.getSample().getUrl() + "</a><br />" +
				"Size: " + post.getSample().getWidth() + "x" + post.getSample().getHeight() + "<br />" +
				"File Size: " + getMB(post.getSample().getSize()) + "<br /><br />" +
				"<strong>File</strong><br />" +
				"URL: <a href=\"" + post.getFile().getUrl() + "\">" + post.getFile().getUrl() + "</a><br />" +
				"Size: " + post.getFile().getWidth() + "x" + post.getFile().getHeight() + "<br />" +
				"File Size: " + getMB(post.getFile().getSize()) + "</p>";
		text.setText(Html.fromHtml(str));
		text.setMovementMethod(LinkMovementMethod.getInstance());
	}

	public String getMB(int bytesize){
		DecimalFormat df = new DecimalFormat("#.#");
		df.setMinimumFractionDigits(2);
		df.setMaximumFractionDigits(2);
		return df.format((float)bytesize / 1024 / 1024) + "MB";
	}

	public void setActionbarIcon(){
		final ActionBar actionBar = getActionBar();
		actionBar.setDisplayUseLogoEnabled(true);
		actionBar.setDisplayShowHomeEnabled(true);
		actionBar.setIcon(R.drawable.ic_action_refresh);
		String path = getApplicationContext().getCacheDir().getAbsolutePath() + "/web_image_cache/" +
				post.getPreview().getUrl().replaceAll("[.:/,%?&=]", "+").replaceAll("[+]+", "+");
		File img = new File(path);
		if(img.exists()){
			Drawable d = Drawable.createFromPath(path);
			actionBar.setIcon(d);
		}else{
			new AsyncTask<Void, Void, Drawable>(){

				@Override
				protected Drawable doInBackground(Void... params){
					try{
						HttpURLConnection connection = (HttpURLConnection)new URL(post.getPreview().getUrl()).openConnection();
						connection.connect();
						InputStream is = connection.getInputStream();
						Bitmap bmp = BitmapFactory.decodeStream(is);
						is.close();
						connection.disconnect();
						return new BitmapDrawable(getResources(), bmp);
					}catch(IOException e){
						return null;
					}
				}

				@Override
				protected void onPostExecute(Drawable d){
					if(d != null)
						actionBar.setIcon(d);
				}
			}.execute();
		}
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu){
		menu.add(Menu.NONE, Menu.FIRST, Menu.NONE, "共有").setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM);
		if(onIntent){
			menu.add(Menu.NONE, Menu.FIRST + 1, Menu.NONE, getString(R.string.open_sample_size)).setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
			menu.add(Menu.NONE, Menu.FIRST + 2, Menu.NONE, getString(R.string.open_full_size)).setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
			menu.add(Menu.NONE, Menu.FIRST + 3, Menu.NONE, getString(R.string.save_full_size)).setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
		}
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item){
		if(item.getItemId() == Menu.FIRST){
			Intent i = new Intent();
			i.setAction(Intent.ACTION_SEND);
			i.setType("text/plain");
			i.putExtra(Intent.EXTRA_TEXT, new Yandere4j().getShareText(post));
			startActivity(i);
		}else if(item.getItemId() == Menu.FIRST + 1){
			Intent i = new Intent(PostDetail.this, ShowImage.class);
			i.putExtra("url", post.getSample().getUrl());
			i.putExtra("filesize", post.getSample().getSize());
			startActivity(i);
		}else if(item.getItemId() == Menu.FIRST + 2){
			Intent i = new Intent(PostDetail.this, ShowImage.class);
			i.putExtra("url", post.getFile().getUrl());
			i.putExtra("filesize", post.getFile().getSize());
			startActivity(i);
		}else if(item.getItemId() == Menu.FIRST + 3){
			((App)getApplicationContext()).saveImage(PostDetail.this, post);
		}
		return super.onOptionsItemSelected(item);
	}
}

File: app/src/main/java/sugtao4423/yandereviewer/PostDetail.kt
package sugtao4423.yandereviewer

import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.Html
import android.text.Spanned
import android.text.format.DateFormat
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import yandere4j.Post
import yandere4j.Yandere4j
import java.io.File
import java.io.IOException
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*
import java.util.regex.Pattern
import javax.net.ssl.HttpsURLConnection

class PostDetail : AppCompatActivity() {

    companion object {
        const val INTENT_EXTRA_POSTDATA = "postdata"
        const val INTENT_EXTRA_ONINTENT = "onIntent"
    }

    private lateinit var post: Post
    private var onIntent = false

    private lateinit var text: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.post_detail)
        text = findViewById(R.id.postDetail_text)
        post = intent.getSerializableExtra(INTENT_EXTRA_POSTDATA) as Post
        onIntent = intent.getBooleanExtra(INTENT_EXTRA_ONINTENT, false)

        val tags = post.tags.joinToString(" ")

        supportActionBar?.title = tags
        setActionbarIcon()

        val urlMatcher = Pattern.compile("http(s)?://([\\w-]+\\.)+[\\w-]+(/[\\w- ./?%&=]*)?").matcher(post.source)
        val source = if (urlMatcher.find()) {
            val pixivMatcher = Pattern.compile(
                    "http(s)?://(i[0-9].pixiv|i.pximg).net/img-original/img/[0-9]{4}/([0-9]{2}/){5}([0-9]+)_p[0-9]+\\..+"
            )
                    .matcher(post.source)
            if (pixivMatcher.find()) {
                "https://www.pixiv.net/artworks/" + pixivMatcher.group(4)
            } else {
                post.source
            }
        } else {
            "https://google.com/search?q=${post.source}"
        }

        val date = DateFormat.getDateFormat(applicationContext).format(post.createdAt)

        val str = "<p><strong>Statistics</strong><br>" +
                "Id: ${post.id}<br>" +
                "Posted: $date " + SimpleDateFormat("HH:mm:ss", Locale.getDefault()).format(post.createdAt) +
                " by ${post.author}<br>" +
                "Tags: $tags<br>" +
                "Size: ${post.file.width}x${post.file.height}<br>" +
                "Source: <a href=\"$source\">${post.source}</a><br>" +
                "Rating: ${post.rating}<br>" +
                "Score: ${post.score}<br><br>" +
                "<strong>Preview</strong><br>" +
                "URL: <a href=\"${post.preview.url}\">${post.preview.url}</a><br>" +
                "Size: ${post.preview.width}x${post.preview.height}<br><br>" +
                "<strong>Sample</strong><br>" +
                "URL: <a href=\"${post.sample.url}\">${post.sample.url}</a><br>" +
                "Size: ${post.sample.width}x${post.sample.height}<br>" +
                "File Size: " + App.getFileMB(post.sample.size) + "<br><br>" +
                "<strong>File</strong><br>" +
                "URL: <a href=\"${post.file.url}\">${post.file.url}</a><br>" +
                "Size: ${post.file.width}x${post.file.height}<br>" +
                "File Size: " + App.getFileMB(post.file.size) + "</p>"
        text.text = fromHtml(str)
        text.movementMethod = MutableLinkMovementMethod().apply {
            setOnUrlClickListener(object : MutableLinkMovementMethod.OnUrlClickListener {
                override fun onUrlClick(widget: TextView, uri: Uri) {
                    ChromeIntent(this@PostDetail, uri)
                }
            })
        }
    }

    @Suppress("DEPRECATION")
    private fun fromHtml(source: String): Spanned {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            Html.fromHtml(source, Html.FROM_HTML_MODE_LEGACY)
        } else {
            Html.fromHtml(source)
        }
    }

    private fun setActionbarIcon() {
        val actionBar = supportActionBar ?: return
        actionBar.apply {
            setDisplayUseLogoEnabled(true)
            setDisplayShowHomeEnabled(true)
            setIcon(R.drawable.ic_action_refresh)
        }

        val path = applicationContext.cacheDir.absolutePath + "/web_image_cache/" +
                post.preview.url.replace(Regex("[.:/,%?&=]"), "+").replace(Regex("[+]+"), "+")
        val img = File(path)
        if (img.exists()) {
            val d = Drawable.createFromPath(path)
            actionBar.setIcon(d)
            return
        }
        CoroutineScope(Dispatchers.Main).launch {
            val result = withContext(Dispatchers.IO) {
                try {
                    val conn = URL(post.preview.url).openConnection() as HttpsURLConnection
                    conn.setRequestProperty("User-Agent", Yandere4j.USER_AGENT)
                    conn.connect()
                    val inputStream = conn.inputStream
                    val bmp = BitmapFactory.decodeStream(inputStream)
                    inputStream.close()
                    conn.disconnect()
                    BitmapDrawable(resources, bmp)
                } catch (e: IOException) {
                    null
                }
            }
            if (result != null) {
                actionBar.setIcon(result)
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menu?.add(Menu.NONE, Menu.FIRST, Menu.NONE, R.string.share)?.setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
        if (onIntent) {
            menu?.add(Menu.NONE, Menu.FIRST + 1, Menu.NONE, R.string.open_sample_size)
                    ?.setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
            menu?.add(Menu.NONE, Menu.FIRST + 2, Menu.NONE, R.string.open_full_size)
                    ?.setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
            menu?.add(Menu.NONE, Menu.FIRST + 3, Menu.NONE, R.string.save_full_size)
                    ?.setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
        }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            Menu.FIRST -> {
                val i = Intent().apply {
                    action = Intent.ACTION_SEND
                    type = "text/plain"
                    putExtra(Intent.EXTRA_TEXT, Yandere4j().getShareText(post))
                }
                startActivity(i)
            }
            Menu.FIRST + 1 -> {
                val i = Intent(this, ShowImage::class.java).apply {
                    putExtra(ShowImage.INTENT_EXTRA_URL, post.sample.url)
                    putExtra(ShowImage.INTENT_EXTRA_FILESIZE, post.sample.size)
                }
                startActivity(i)
            }
            Menu.FIRST + 2 -> {
                val i = Intent(this, ShowImage::class.java).apply {
                    putExtra(ShowImage.INTENT_EXTRA_URL, post.file.url)
                    putExtra(ShowImage.INTENT_EXTRA_FILESIZE, post.file.size)
                }
                startActivity(i)
            }
            Menu.FIRST + 3 -> {
                (applicationContext as App).saveImage(this, post)
            }
        }
        return super.onOptionsItemSelected(item)
    }

}